server/controllers/ArticleController.js

```diff
@@ -7,21 +7,167 @@ import models from '../models';
 import numberOfArticles from '../helpers/numberOfArticles';
 
 const {
-  Article, User, ArticleLike, Tag, ArticleTag, Comment
+  Article, User, ArticleLike, Tag, ArticleTag, Comment, Bookmark
 } = models;
 const { Op } = Sequelize;
 /**
  * @class ArticleController
  * @description Article related Operations
  */
 class ArticleController {
-/**
+  /**
+ * @description
+ * @param {object} req - request object
+ * @param {object} res - response object
+ * @returns {object} - bookmarks an Article
+ */
+  static bookmarkArticle(req, res) {
+    const userId = req.userData.id;
+    const { articleId } = req.params;
+    Bookmark.findOrCreate({
+      where: {
+        userId,
+        articleId
+      }
+    })
+      .spread((returnedBookmarkedArticle, status) => {
+        if (!status) {
+          return res.status(403).json({
+            status: 'failure',
+            errors: {
+              message: ['you have already bookmarked this article']
+            }
+          });
+        }
+        if (returnedBookmarkedArticle) {
+          req.params.slug = articleId;
+          ArticleController
+            .getAnArticle(req, res, 'Article successfully bookmarked');
+        }
+      });
+  }
+
+  /**
+ * @description
+ * @param {object} req - request object
+ * @param {object} res - response object
+ * @returns {object} - bookmarks an Article
+ */
+  static unbookmarkArticle(req, res) {
+    const userId = req.userData.id;
+    const { articleId } = req.params;
+    Bookmark.findOne({
+      where: {
+        userId,
+        articleId
+      },
+      include: [{
+        model: Article,
+        as: 'bookmarkedArticles'
+      }]
+    })
+      .then((returnedBookmarkedArticle) => {
+        if (!returnedBookmarkedArticle) {
+          return res.status(404).json({
+            status: 'failure',
+            errors: {
+              message: ['you didnt bookmark this article']
+            }
+          });
+        }
+        if (returnedBookmarkedArticle) {
+          Bookmark.destroy({
+            where: {
+              userId,
+              articleId
+            }
+          })
+            .then(() => {
+              req.params.slug = articleId;
+              ArticleController
+                .getAnArticle(
+                  req, res, 'Article successfully unbookmarked'
+                );
+            });
+        }
+      });
+  }
+
+  /**
+ * @description gets bookmarked Article
+ * @param {object} req - request object
+ * @param {object} res - response object
+ * @returns {object} - json response
+ */
+  static getbookmarkedArticle(req, res) {
+    const userId = req.userData.id;
+    const { articleId } = req.params;
```

Comment:
I do not think this part is needed. We can get all bookmarked articles by the userId from the Bookmarks table. We do not need the articleId. Can you colour this maybe there is something I do not see.

Reply:
That's fine. On the other hand, it means a user cannot fetch one of his bookmarked articles by its id.

Reply:
A user can fetch his/her article by id via the fetch article by id route